client/driver/env/env_test.go

```diff
@@ -318,6 +321,139 @@ func TestEnvironment_AsList_Old(t *testing.T) {
 	require.Equal(t, exp, act)
 }
 
+func TestEnvironment_AllValues(t *testing.T) {
+	t.Parallel()
+
+	n := mock.Node()
+	n.Meta = map[string]string{
+		"metaKey":           "metaVal",
+		"nested.meta.key":   "a",
+		"invalid...metakey": "b",
+	}
+	a := mock.Alloc()
+	a.AllocatedResources.Tasks["web"].Networks[0] = &structs.NetworkResource{
+		Device:        "eth0",
+		IP:            "127.0.0.1",
+		ReservedPorts: []structs.Port{{Label: "https", Value: 8080}},
+		MBits:         50,
+		DynamicPorts:  []structs.Port{{Label: "http", Value: 80}},
+	}
+	a.AllocatedResources.Tasks["ssh"] = &structs.AllocatedTaskResources{
+		Networks: []*structs.NetworkResource{
+			{
+				Device: "eth0",
+				IP:     "192.168.0.100",
+				MBits:  50,
+				ReservedPorts: []structs.Port{
+					{Label: "ssh", Value: 22},
+					{Label: "other", Value: 1234},
+				},
+			},
+		},
+	}
+	task := a.Job.TaskGroups[0].Tasks[0]
+	task.Env = map[string]string{
+		"taskEnvKey":        "taskEnvVal",
+		"nested.task.key":   "x",
+		"invalid...taskkey": "y",
+	}
+	env := NewBuilder(n, a, task, "global").SetDriverNetwork(
+		&cstructs.DriverNetwork{PortMap: map[string]int{"https": 443}},
+	)
+
+	values, errs, err := env.Build().AllValues()
+	require.NoError(t, err)
+
+	// Assert the keys we couldn't nest were reported
+	require.Len(t, errs, 2)
+	require.Contains(t, errs, "invalid...taskkey")
+	require.Contains(t, errs, "meta.invalid...metakey")
+
+	exp := map[string]string{
+		// Node
+		"node.unique.id":          n.ID,
+		"node.region":             "global",
+		"node.datacenter":         n.Datacenter,
+		"node.unique.name":        n.Name,
+		"node.class":              n.NodeClass,
+		"meta.metaKey":            "metaVal",
+		"attr.arch":               "x86",
+		"attr.driver.exec":        "1",
+		"attr.driver.mock_driver": "1",
+		"attr.kernel.name":        "linux",
+		"attr.nomad.version":      "0.5.0",
+
+		// 0.9 style meta and attr
+		"node.meta.metaKey":            "metaVal",
```

Comment:
What if the node meta is invalid since we don't control those either

Reply:
Hm, right now it's just logged and dropped. We could easily expose them all inside the `env` map or create new map for node and its children. Looking back at our doc it's not clear what our plan was.

Reply:
Resolving and fixing the following in a followup PR:

* Add Upgrade Guide entry
* Validate Node `meta` on agent startup